components/net/http_loader.rs

```diff
@@ -316,30 +286,41 @@ impl Read for StreamedResponse {
 }
 
 impl StreamedResponse {
-    fn from_http_response(response: WrappedHttpResponse) -> io::Result<StreamedResponse> {
-        let decoder = match response.content_encoding() {
-            Some(Encoding::Gzip) => {
-                Decoder::Gzip(GzDecoder::new(response)?)
-            }
-            Some(Encoding::Deflate) => {
-                Decoder::Deflate(DeflateDecoder::new(response))
-            }
-            Some(Encoding::EncodingExt(ref ext)) if ext == "br" => {
-                Decoder::Brotli(Decompressor::new(response, 1024))
+    fn from_http_response(response: HyperResponse) -> io::Result<StreamedResponse> {
+        let mut encode_type = "";
+        let decoder;
+        if let Some(encoding) = response.headers.get::<ContentEncoding>() {
```

Comment:
Let's do this instead:
```rust
let decoder = if let Some(encoding) = response.headers.get::<ContentEncoding>() {
    if encoding.contains(&Encoding::Gzip) {
        Decoder::Gzip(GzDecoder::new(response)?)
    } else if encoding.contains(...) {
        ...
    }
} else {
    Decoder::Plain(response)
};
```

Reply:
@jdm 
this will cause borrow issue
I don't know how to solve it
```
error[E0505]: cannot move out of `response` because it is borrowed
   --> components/net/http_loader.rs:297:61
    |
291 |         if let Some(encoding) = response.headers.get::<ContentEncoding>(){
    |                                 ---------------- borrow of `response.headers` occurs here
...
297 |                 decoder = Decoder::Brotli(Decompressor::new(response, 1024));
    |                                                             ^^^^^^^^ move out of `response` occurs here
```

Reply:
Use `ref encoding`.

Reply:
Sorry for not familiar with Rust...

@jdm Do you mean this?
```rust
impl StreamedResponse {
    fn from_http_response(response: HyperResponse) -> io::Result<StreamedResponse> {
        let decoder;
        if let Some(ref encoding) = response.headers.get::<ContentEncoding>() {
            if encoding.contains(&Encoding::Gzip) {
                decoder = Decoder::Gzip(GzDecoder::new(response)?);
            } else if encoding.contains(&Encoding::Deflate) {
                decoder = Decoder::Deflate(DeflateDecoder::new(response));
            } else if encoding.contains(&Encoding::EncodingExt("br".to_owned())) {
                decoder = Decoder::Brotli(Decompressor::new(response, 1024));
            }
        } else {
            decoder = Decoder::Plain(response);
        }
        Ok(StreamedResponse { decoder: decoder })
    }
}
```
it will get
```shell
error[E0505]: cannot move out of `response` because it is borrowed
   --> components/net/http_loader.rs:293:56
    |
291 |         if let Some(ref encoding) = response.headers.get::<ContentEncoding>() {
    |                                     ---------------- borrow of `response.headers` occurs here
292 |             if encoding.contains(&Encoding::Gzip) {
293 |                 decoder = Decoder::Gzip(GzDecoder::new(response)?);
    |                                                        ^^^^^^^^ move out of `response` occurs here

error[E0505]: cannot move out of `response` because it is borrowed
   --> components/net/http_loader.rs:295:64
    |
291 |         if let Some(ref encoding) = response.headers.get::<ContentEncoding>() {
    |                                     ---------------- borrow of `response.headers` occurs here
...
295 |                 decoder = Decoder::Deflate(DeflateDecoder::new(response));
    |                                                                ^^^^^^^^ move out of `response` occurs here

error[E0505]: cannot move out of `response` because it is borrowed
   --> components/net/http_loader.rs:297:61
    |
291 |         if let Some(ref encoding) = response.headers.get::<ContentEncoding>() {
    |                                     ---------------- borrow of `response.headers` occurs here
...
297 |                 decoder = Decoder::Brotli(Decompressor::new(response, 1024));
    |                                                             ^^^^^^^^ move out of `response` occurs here

error[E0505]: cannot move out of `response` because it is borrowed
   --> components/net/http_loader.rs:300:38
    |
291 |         if let Some(ref encoding) = response.headers.get::<ContentEncoding>() {
    |                                     ---------------- borrow of `response.headers` occurs here
...
300 |             decoder = Decoder::Plain(response);
    |                                      ^^^^^^^^ move out of `response` occurs here

error[E0381]: use of possibly uninitialized variable: `decoder`
   --> components/net/http_loader.rs:302:40
    |
302 |         Ok(StreamedResponse { decoder: decoder })
    |                                        ^^^^^^^ use of possibly uninitialized `decoder`


```

Reply:
The original implement avoid call `response` in a `response` calling.
But this method will encounter this problem.

Reply:
Ah, I see. In that case, instead of using strings let's use a simple `enum DecoderType { Gzip, Deflate, Brotli, Plain }`.

Reply:
I think calling `.clone()` on `response.headers.get::<ContentEncoding>()` should solve the problem.

Reply:
True, but no need to clone the whole header.